saleor/graphql/core/connection.py

```diff
@@ -1,11 +1,226 @@
+    if not isinstance(values, Iterable):
+        values = [values]
+    values = [value if value is None else str(value) for value in values]
+    return base64(json.dumps(values))
+
+
+def from_global_cursor(cursor) -> List[str]:
+    values = unbase64(cursor)
+    return json.loads(values)
+
+
+def get_field_value(instance: DjangoModel, field_name: str):
+    """Get field value for given field in filter format 'field__foreign_key_field'."""
+    field_path = field_name.split("__")
+    attr = instance
+    for elem in field_path:
+        attr = getattr(attr, elem)
+
+    if callable(attr):
+        return "%s" % attr()
+    return attr
+
+
+def _prepare_filter_expression(
+    field_name: str,
+    index: int,
+    cursor: List[str],
+    sorting_fields: List[str],
+    sorting_direction: str,
+) -> Tuple[Q, Dict[str, Union[str, bool]]]:
+
+    field_expression: Dict[str, Union[str, bool]] = {}
+    extra_expression = Q()
+    for cursor_id, cursor_value in enumerate(cursor[:index]):
+        field_expression[sorting_fields[cursor_id]] = cursor_value
+
+    if sorting_direction == "gt":
+        extra_expression |= Q(**{f"{field_name}__{sorting_direction}": cursor[index]})
+        extra_expression |= Q(**{f"{field_name}__isnull": True})
+    elif cursor[index] is not None:
+        field_expression[f"{field_name}__{sorting_direction}"] = cursor[index]
+    else:
+        field_expression[f"{field_name}__isnull"] = False
+
+    return extra_expression, field_expression
+
+
+def _prepare_filter(
+    cursor: List[str], sorting_fields: List[str], sorting_direction: str
+) -> Q:
+    """Create filter arguments based on sorting fields.
+
+    :param cursor: list of values that are passed from page_info, used for filtering.
+    :param sorting_fields: list of fields that were used for sorting.
+    :param sorting_direction: keyword direction ('lt', gt').
+    :return: Q() in following format
+        (OR: ('first_field__gt', 'first_value_form_cursor'),
+            (AND: ('second_field__gt', 'second_value_form_cursor'),
+                ('first_field', 'first_value_form_cursor')),
+            (AND: ('third_field__gt', 'third_value_form_cursor'),
+                ('second_field', 'second_value_form_cursor'),
+                ('first_field', 'first_value_form_cursor'))
+        )
+    """
+    filter_kwargs = Q()
+    for index, field_name in enumerate(sorting_fields):
+        if cursor[index] is None and sorting_direction == "gt":
+            continue
+
+        extra_expression, field_expression = _prepare_filter_expression(
+            field_name, index, cursor, sorting_fields, sorting_direction
+        )
+        filter_kwargs |= Q(extra_expression, **field_expression)
+
+    return filter_kwargs
+
+
+def _validate_connection_args(args):
+    first = args.get("first")
+    last = args.get("last")
+
+    if first and not (isinstance(first, int) and first > 0):
+        raise ValueError('Argument "first" must be a non-negative integer.')
```

Comment:
I suggest that we change these `ValueError`s to `GraphQLError`. This would be consistent with other errors returned by Graphene when passing an invalid query.